File: Python/chat_server.py
# chat_server_async.py
import asyncio

clients: set[asyncio.StreamWriter] = set()
clients_lock = asyncio.Lock()

async def broadcast(message: str, sender: asyncio.StreamWriter | None = None):
    data = (message + "\n").encode()
    async with clients_lock:
        dead = []
        for w in clients:
            if w is sender:
                continue
            try:
                w.write(data)
                await w.drain()
            except Exception:
                dead.append(w)
        for w in dead:
            clients.discard(w)

async def handle_client(reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
    addr = writer.get_extra_info("peername")
    name = f"{addr[0]}:{addr[1]}"
    async with clients_lock:
        clients.add(writer)

    writer.write(f"Welcome {name}! Type and press ENTER.\n".encode())
    await writer.drain()
    await broadcast(f"[+] {name} joined", sender=None)

    try:
        while True:
            line = await reader.readline()
            if not line:
                break
            msg = line.decode(errors="ignore").rstrip("\n")
            if msg.strip().lower() == "/quit":
                break
            await broadcast(f"[{name}] {msg}", sender=writer)
    except asyncio.CancelledError:
        raise
    except Exception as e:
        print(f"[!] {name} error: {e}")
    finally:
        async with clients_lock:
            clients.discard(writer)
        await broadcast(f"[-] {name} left", sender=None)
        writer.close()
        await writer.wait_closed()

async def main(host="0.0.0.0", port=9001):
    server = await asyncio.start_server(handle_client, host, port)
    addrs = ", ".join(str(sock.getsockname()) for sock in server.sockets)
    print(f"Chat server listening on {addrs}")
    async with server:
        await server.serve_forever()

if __name__ == "__main__":
    asyncio.run(main())

File: Go/chat_server.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"net"
	"strings"
	"sync"
)

const (
	HOST = "0.0.0.0"
	PORT = "9001"
)

type Client struct {
	conn net.Conn
	name string
	out  chan string
}

var (
	mu      sync.Mutex
	clients = make(map[net.Conn]*Client)
)

func main() {
	addr := net.JoinHostPort(HOST, PORT)
	ln, err := net.Listen("tcp", addr)
	if err != nil {
		log.Fatalf("[Server error] %v", err)
	}
	defer ln.Close()

	log.Printf("Chat server listening on %s\n", addr)

	for {
		conn, err := ln.Accept()
		if err != nil {
			log.Printf("[Accept error] %v\n", err)
			continue
		}
		go handleClient(conn)
	}
}

func handleClient(conn net.Conn) {
	name := conn.RemoteAddr().String()

	c := &Client{
		conn: conn,
		name: name,
		out:  make(chan string, 32),
	}

	mu.Lock()
	clients[conn] = c
	count := len(clients)
	mu.Unlock()

	log.Printf("[+] %s joined (%d clients)\n", name, count)

	// Writer goroutine
	go func() {
		for msg := range c.out {
			_, _ = conn.Write([]byte(ensureNL(msg)))
		}
	}()

	// Welcome + broadcast join
	c.out <- fmt.Sprintf("Welcome %s! Type message. Use /quit to exit.", name)
	broadcast(fmt.Sprintf("[+] %s joined", name), conn)

	reader := bufio.NewReader(conn)
	for {
		line, err := reader.ReadString('\n') // newline framing như JS
		if err != nil {
			break
		}

		line = strings.TrimRight(line, "\r\n")
		msg := strings.TrimSpace(line)
		if msg == "" {
			continue
		}

		if strings.EqualFold(msg, "/quit") {
			// giống JS: socket.end("Bye!\n")
			c.out <- "Bye!"
			break
		}

		log.Printf("[%s] %s\n", name, msg)
		broadcast(fmt.Sprintf("[%s] %s", name, msg), conn)
	}

	// Cleanup
	mu.Lock()
	delete(clients, conn)
	count = len(clients)
	mu.Unlock()

	log.Printf("[-] %s left (%d clients)\n", name, count)
	broadcast(fmt.Sprintf("[-] %s left", name), nil)

	close(c.out)
	_ = conn.Close()
}

func broadcast(message string, except net.Conn) {
	data := ensureNL(message)
	mu.Lock()
	defer mu.Unlock()
	for conn, c := range clients {
		if except != nil && conn == except {
			continue
		}
		select {
		case c.out <- data:
		default:
			// nếu client chậm, bỏ qua để server không bị kẹt
		}
	}
}

func ensureNL(s string) string {
	if strings.HasSuffix(s, "\n") {
		return s
	}
	return s + "\n"
}
